docs: Improve pointing docs (#2703)
zmkfirmware/zmk: docs/docs/development/hardware-integration/pointing.mdx

> ---

Author: I added a new section on configuration, seems useful to have it to me.

toc_max_heading_level: 2
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

ZMK's pointing device support builds upon the Zephyr [input API](https://docs.zephyrproject.org/3.5.0/services/input/index.html) to offer pointing/mouse functionality with various hardware.
A limited number of input drivers are available in the Zephyr version currently used by ZMK, but additional drivers can be found in [external modules](../../features/modules.mdx) for a variety of hardware.

Pointing devices are also supported on split peripherals, with some additional configuration using the [input split device](../../config/pointing.md#input-split).
The configuration details will depend on if you are adding a pointing device to a [split peripheral](../../features/split-keyboards.md#central-and-peripheral-roles) as opposed to a unibody keyboard or split central part.

export const SplitTabs = (props) => (
  <Tabs
    groupId="part-type"
    defaultValue="central"
    values={[
      { label: "Unibody or Split Central", value: "central" },
      { label: "Split Peripheral", value: "peripheral" },
    ]}
  >

    {/* eslint-disable-next-line */}
    {props.children}

  </Tabs>

);

> 

Author: As long as the selections are synced, I think having separate tabs isn't bad. I find it confusing when you select a tab and it affects things way down the page, and you can't switch back without scrolling to the top.

> 

Reviewer: How about just moving the first tab with the trick and leaving the second untouched?

> 

Reviewer: Happy with the tabs now as-is.

See [input listener configuration](../../config/pointing.md#input-listener) for the full details.

<SplitTabs>
  <TabItem value="central">

For example, to add a listener for the above device, add to your `.overlay`/`.dts` file for the unibody or central part a node like the following:

```dts title="<central>.overlay"
/ {
    glidepoint_listener {
        compatible = "zmk,input-listener";
        device = <&glidepoint>;
    };
};
```

  </TabItem>
  <TabItem value="peripheral">

### Shared Configuration

When a pointing device is on a peripheral, both peripheral and central make use of a `zmk,input-split` device, which functions differently depending on where it is used.
To avoid duplicating work, this node can be defined in a common `.dtsi` file that is included into both central and peripheral `.overlay`/`.dts` files.

All split pointers are identified using a unique integer value, which is specified using the `reg` property and in the `@#` suffix for the node. If adding multiple peripheral pointers, be sure that each is given a unique identifier.

> Second, the input listener for the central side is added here, but disabled, so that keymaps (which are included for central and peripheral builds) can reference the listener to add input processors without issue.

Author: I expanded it a bit, and added a note under node definition that describes how the user uses the node reference.

> Second, the input listener for the central side is added here, but disabled, so that keymaps (which are included for central and peripheral builds) can reference the listener to add input processors without issue.

Author: Also added a new tab for split centrals vs. unibody, should be correct from build experiments I did, though I don't have hardware to test on.

};
```

The specifics of the properties required to set for a given driver will vary; always consult the devicetree bindings file for the specific driver to see what properties can be set.

## Listener and Input Split Device

Every input device needs an associated listener added that listens for events from the device and processes them before sending the events to the host using a HID mouse report.
See [input listener configuration](../../config/pointing.md#input-listener) for the full details.

If your pointing device is on a split peripheral part, you also need to define and use an input split device on all keyboard parts.

<SplitTabs>
  <TabItem value="unibody">

To add a listener for the above device, add to your `.overlay`/`.dts` file for the keyboard a node like the following:

```dts title="<keyboard>.overlay"
/ {
    glidepoint_listener {
        compatible = "zmk,input-listener";
        device = <&glidepoint>;
    };
};
```

  </TabItem>
  <TabItem value="central">

### Shared Configuration

> 

Reviewer: I don't think I'm a fan of suggesting the use of a separate .dtsi file for the pointing stuff, it doesn't seem long enough to be worthwhile and putting it into the standard keyboard.dtsi avoids any "forgot to include" issues.

> The input listener that is used by the central side is added in a shared in a common `.dtsi` file that is included into both central and peripheral `.overlay`/`.dts` files, but disabled by default.

Reviewer: "in a shared in a common"

};
```

  </TabItem>
  <TabItem value="central">

### Shared Configuration

> The input listener that is used by the central side is added in the shared `.dtsi` file that is included into both central and peripheral `.overlay`/`.dts` files, but disabled by default.
> This is so that keymaps (which are included for both central and peripheral builds) can reference the listener to add input processors without failing with an undefined reference error.

Reviewer: ```suggestion
The input listener that is used by the central side is added to the shared `.dtsi` file that is included into both central and peripheral `.overlay`/`.dts` files.

The input listener is disabled by default, and will be enabled later in the central part's overlay. This is so that keymaps (which are included for both central and peripheral builds) can reference the listener to add input processors without failing with an undefined reference error.
```

```dts title="<keyboard>.dtsi"
/ {
    glidepoint_listener: glidepoint_listener {
        compatible = "zmk,input-listener";
        status = "disabled";
    };
};
```

> If users want to add [input processors](#input-processors) to the listener node, they will use the node reference `&glidepoint_listener` in their keymaps.

Reviewer: ```suggestion
If users want to add [input processors](#input-processors) to the listener node, they will use the corresponding node reference (here it is `&glidepoint_listener`) in their keymaps.
```

> Users will use this node reference `&glidepoint_listener` in their keymaps if they want to add [input processors](#input-processors) to it.

Reviewer: This sentence reads oddly

Author: I reworded it, let me know if you can think of a clearer rewording.

### Central Configuration

In the central, we do the following:

- Include the shared .dtsi file.
- Enable the input listener that is created in our shared file.
- Assign the pointing device node to the listener.

```dts title="<central>.overlay"
// Pull in the shared configuration
#include "<keyboard>-pointing.dtsi"

> // Node added in Input Device section above

Reviewer: ```suggestion
// Node from the previous Input Device section
```

};
```

> If users want to add [input processors](#input-processors) to the listener node, they will use the node reference `&glidepoint_listener` in their keymaps.

Reviewer: ```suggestion
If users want to add [input processors](#input-processors) to the listener node, they will use the corresponding node reference (here it is `&glidepoint_listener`) in their keymaps.
```

### Peripheral Configuration

In the peripheral .overlay/.dts file, we do the following:

- Include the shared .dtsi file.
- Update the input split with a reference to the new device node that should be proxied.

```dts title="<peripheral>.overlay"
// Pull in the shared configuration
#include "<keyboard>-pointing.dtsi"

> // Node added in Input Device section above

Reviewer: ```suggestion
// Node from the previous Input Device section
```

```dts title="<central>.overlay"
#include "<keyboard>-pointing.dtsi"

&glidepoint_listener {
    status = "okay";
};
```

  </TabItem>
</SplitTabs>

## Input Processors

Some physical pointing devices may be generating input events that need adjustment before being sent to hosts.
For example a trackpad might be integrated into a keyboard rotated 90° and need the X/Y data adjusted appropriately.
This can be accomplished with [input processors](../../keymaps/input-processors/index.md).
As an example, you could enhance the listener defined in the previous section with an input processor that inverts and swaps the X/Y axes:

```dts
#include <dt-bindings/zmk/input_transform.h>

/ {
    glidepoint_listener {
        compatible = "zmk,input-listener";
        device = <&glidepoint>;
        input-processors = <&zip_xy_transform (INPUT_TRANSFORM_XY_SWAP | INPUT_TRANSFORM_X_INVERT | INPUT_TRANSFORM_Y_INVERT)>;
    };
};
```

## Configuration Setting

If your keyboard hardware includes a pointing device by default, you can enable the [`ZMK_POINTING` config](../../config/pointing.md#general) in your keyboard definition.

> You can do that in your [`Kconfig.defconfig` file](new-shield.mdx#kconfigdefconfig), where you can also enable the config for the communication protocol used by the pointing device:

Reviewer: ```suggestion
You can do that in your [`Kconfig.defconfig` file](new-shield.mdx#kconfigdefconfig), where you can also enable the config for the communication protocol (e.g. SPI, I2C) used by the pointing device:
```
Maybe also link the corresponding protocol to the corresponding Zephyr Kconfig flag docs?

<SplitTabs>
  <TabItem value="unibody">

```kconfig title="Kconfig.defconfig"
if SHIELD_MY_KEYBOARD

# Other keyboard settings

config ZMK_POINTING
    default y

> 

Reviewer: I'd add some extra comments relating the SPI config flag to the pointing device here.

Author: Can you elaborate what kind of comment you are thinking of?

Reviewer: Something like "If the above pointing device uses SPI", just to help avoid blind copypaste syndrome a bit more
